designs/node-messaging-api.md

```diff
@@ -41,17 +41,81 @@ This simple system has a few limitations that we want to address:
  3. due to 1 & 2, we cannot build any timeout feature into the node
  4. we know of a use case where an embedder of node-red requires nodes to be strictly one-in, one-out, and that should be policed by the runtime. Putting aside the specifics, we cannot provide any such mode of operation with the current model
 
-> (HN): According to the discussion held on May 17th,
-> the correlation information among an input message and output messages are expected to be used for logging purpose, and
-> and a output message correlate with the latest input message received.
-
 This design note explores how this mechanism could be updated to satisfy these limitations.
 
-The original draft of this node set out three options. Through the process of writing the options, I've chosen my preferred approach. I have left the other two options at the bottom of the note for reference.
+> This design has changed a few times. Having settled on one approach, concerns
+> were raised that it was too much of a change from the existing API to be
+> practical to adopt. This latest design reflects on that and tries to find
+> a more modest approach that satisfies as much of the requirements as possible.
+
+The primary use case for this is for a user to be able to create a Flow that is
+notified when a message is successfully processed at a critical point of a flow.
+For example, the `Email Out` node has published its message.
+
+### `Node.done([statusObject], msg)`
+
+A new function is added to the `Node` object that can be used to indicate a node
+has finished processing a message.
+
+The function takes two arguments:
+
+ - `statusObject` - an _optional_ object used to update the Node's status.
+ - `msg` - the message being marked as complete.
+
+Calling this function will trigger any in-scope `Status` nodes. They will emit
+the `msg` passed to the function. If `statusObject` is provided, `msg.status` will
+be set to its value. In any case, the `msg` will have `msg.status.complete` set
+to `true`.
+
+The Status event will _only_ be passed to the editor over comms if `statusObject`
+has been provided.
+
+This function should be the last thing a node does after handling a message. It
+should only be called once per message received.
+
+```
```

Comment:
So this example doesn't go to the editor ? - so it just goes to the status node ? Can the statusObject be just text ? Are there examples of statusObject just for clarity ?

Reply:
Calling `node.done(msg)` will trigger any Status nodes, but will *not* push anything to the editor.

The `statusObject` will be consistent with what you can pass `node.status(statusObject)` - so either an object with `text,fill,shape` properties, or just a string as a convenience for `{text:"my status"}`. In either case, it will will be used as the base status object that has `complete:true` added before being passed to Status nodes *and* will push the status to the editor.